File: frontend/assets/js/formatters.js
"""
Data formatting utilities
"""

from typing import Dict, Any


def formatFileSize(bytes: int) -> str:
    """Format file size in bytes to human readable format"""
    if bytes == 0:
        return "0 Bytes"
    
    import math
    k = 1024
    sizes = ["Bytes", "KB", "MB", "GB"]
    i = int(math.floor(math.log(bytes) / math.log(k)))
    
    return f"{math.floor((bytes / math.pow(k, i)) * 100) / 100} {sizes[i]}"


def formatScanResult(scanResult: Dict[str, Any]) -> str:
    """Format OCR scan result to HTML"""
    if not scanResult:
        return "<p>Không có dữ liệu</p>"

    # Kiểm tra nếu là căn cước công dân Lào thì sử dụng format chuyên dụng
    if scanResult.get('document_type') == 'lao_cccd':
        return formatLaoCCCDResult(scanResult)

    # Format chung cho các loại tài liệu khác
    html = '<div class="scan-result">'

    if scanResult.get('text'):
        html += f'<h4><i class="fas fa-text-width"></i> Văn bản trích xuất:</h4>'
        html += f'<p class="extracted-text">{scanResult["text"]}</p>'

    if scanResult.get('document_type'):
        html += f'<h4><i class="fas fa-id-card"></i> Loại tài liệu:</h4>'
        html += f'<p class="document-type">{scanResult["document_type"]}</p>'

    if scanResult.get('display_name'):
        html += f'<h4><i class="fas fa-user"></i> Tên hiển thị:</h4>'
        html += f'<p class="display-name">{scanResult["display_name"]}</p>'

    # Add other fields
    excluded_fields = ["text", "document_type", "display_name", "img_base64"]
    for key, value in scanResult.items():
        if key not in excluded_fields:
            html += f'<h4><i class="fas fa-info-circle"></i> {key}:</h4>'
            html += f'<p>{value}</p>'

    html += "</div>"
    return html


def formatLaoCCCDResult(ocrResult: Dict[str, Any]) -> str:
    """Format OCR result for Lao CCCD to beautiful HTML display"""
    if not ocrResult:
        return "<p>Không có dữ liệu căn cước</p>"

    html = '<div class="lao-cccd-result">'
    
    # Header với tên tài liệu
    html += '<div class="cccd-header">'
    html += f'<h3><i class="fas fa-id-card"></i> {ocrResult.get("display_name", "Căn cước công dân Lào")}</h3>'
    html += '</div>'
    
    # Thông tin cá nhân chính
    fields = ocrResult.get('fields', {})
    if fields:
        html += '<div class="cccd-section">'
        html += '<h4><i class="fas fa-user"></i> Thông tin cá nhân</h4>'
        html += '<div class="cccd-info-grid">'
        
        # Số căn cước
        if fields.get('id_number'):
            html += '<div class="info-item">'
            html += '<span class="info-label"><i class="fas fa-hashtag"></i> Số căn cước:</span>'
            html += f'<span class="info-value">{fields["id_number"]}</span>'
            html += '</div>'
        
        # Họ tên
        if fields.get('fullname'):
            html += '<div class="info-item">'
            html += '<span class="info-label"><i class="fas fa-signature"></i> Họ và tên:</span>'
            html += f'<span class="info-value">{fields["fullname"]}</span>'
            html += '</div>'
        
        # Ngày sinh
        if fields.get('dob'):
            html += '<div class="info-item">'
            html += '<span class="info-label"><i class="fas fa-birthday-cake"></i> Ngày sinh:</span>'
            html += f'<span class="info-value">{fields["dob"]}</span>'
            html += '</div>'
        
        # Quốc tịch
        if fields.get('nationality'):
            html += '<div class="info-item">'
            html += '<span class="info-label"><i class="fas fa-flag"></i> Quốc tịch:</span>'
            html += f'<span class="info-value">{fields["nationality"]}</span>'
            html += '</div>'
        
        # Dân tộc
        if fields.get('ethnicity'):
            html += '<div class="info-item">'
            html += '<span class="info-label"><i class="fas fa-users"></i> Dân tộc:</span>'
            html += f'<span class="info-value">{fields["ethnicity"]}</span>'
            html += '</div>'
        
        html += '</div>'
        html += '</div>'
        
        # Thông tin địa chỉ
        address = fields.get('address', {})
        if address:
            html += '<div class="cccd-section">'
            html += '<h4><i class="fas fa-map-marker-alt"></i> Địa chỉ</h4>'
            html += '<div class="cccd-info-grid">'
            
            # Địa chỉ đầy đủ
            if address.get('address'):
                html += '<div class="info-item full-width">'
                html += '<span class="info-label"><i class="fas fa-home"></i> Địa chỉ:</span>'
                html += f'<span class="info-value">{address["address"]}</span>'
                html += '</div>'
            
            # Chi tiết địa chỉ
            children = address.get('childrent', {})
            if children:
                if children.get('address_village'):
                    html += '<div class="info-item">'
                    html += '<span class="info-label"><i class="fas fa-building"></i> Thôn/Bản:</span>'
                    html += f'<span class="info-value">{children["address_village"]}</span>'
                    html += '</div>'
                
                if children.get('address_district'):
                    html += '<div class="info-item">'
                    html += '<span class="info-label"><i class="fas fa-map"></i> Huyện:</span>'
                    html += f'<span class="info-value">{children["address_district"]}</span>'
                    html += '</div>'
                
                if children.get('address_province'):
                    html += '<div class="info-item">'
                    html += '<span class="info-label"><i class="fas fa-city"></i> Tỉnh:</span>'
                    html += f'<span class="info-value">{children["address_province"]}</span>'
                    html += '</div>'
            
            html += '</div>'
            html += '</div>'
        
        # Thông tin thời hạn
        html += '<div class="cccd-section">'
        html += '<h4><i class="fas fa-calendar-alt"></i> Thông tin thời hạn</h4>'
        html += '<div class="cccd-info-grid">'
        
        if fields.get('issue_date'):
            html += '<div class="info-item">'
            html += '<span class="info-label"><i class="fas fa-calendar-plus"></i> Ngày cấp:</span>'
            html += f'<span class="info-value">{fields["issue_date"]}</span>'
            html += '</div>'
        
        if fields.get('expiry_date'):
            html += '<div class="info-item">'
            html += '<span class="info-label"><i class="fas fa-calendar-times"></i> Ngày hết hạn:</span>'
            html += f'<span class="info-value">{fields["expiry_date"]}</span>'
            html += '</div>'
        
        html += '</div>'
        html += '</div>'
    
    # Hiển thị ảnh nếu có
    if ocrResult.get('img_url'):
        html += '<div class="cccd-section">'
        html += '<h4><i class="fas fa-image"></i> Ảnh căn cước</h4>'
        html += '<div class="cccd-image-container">'
        html += f'<img src="{ocrResult["img_url"]}" alt="Ảnh căn cước công dân" class="cccd-image" />'
        html += '</div>'
        html += '</div>'
    
    # Văn bản gốc (nếu cần debug)
    if ocrResult.get('text'):
        html += '<div class="cccd-section">'
        html += '<h4><i class="fas fa-file-text"></i> Văn bản trích xuất</h4>'
        html += '<div class="cccd-text-container">'
        html += f'<pre class="cccd-text">{ocrResult["text"]}</pre>'
        html += '</div>'
        html += '</div>'
    
    html += '</div>'
    return html


def formatVerifyResult(verifyResult: Dict[str, Any]) -> str:
    """Format face verification result to HTML"""
    if not verifyResult:
        return "<p>Không có dữ liệu xác thực</p>"

    html = '<div class="verify-result">'

    # Handle new WebSocket API format
    if verifyResult.get('status') == "success":
        html += '<h4><i class="fas fa-check-circle"></i> Trạng thái:</h4>'
        html += '<p class="match-success">Xác thực thành công</p>'

        if verifyResult.get('same_person') is not None:
            matchStatus = "Cùng một người" if verifyResult['same_person'] else "Không cùng người"
            matchClass = "match-success" if verifyResult['same_person'] else "match-fail"
            html += '<h4><i class="fas fa-user-check"></i> Kết quả:</h4>'
            html += f'<p class="{matchClass}">{matchStatus}</p>'

        if verifyResult.get('similarity') is not None:
            html += '<h4><i class="fas fa-percentage"></i> Độ tương đồng:</h4>'
            html += f'<p class="match-score">{(verifyResult["similarity"] * 100):.2f}%</p>'

        if verifyResult.get('msg'):
            html += '<h4><i class="fas fa-comment"></i> Thông báo:</h4>'
            html += f'<p>{verifyResult["msg"]}</p>'
    else:
        html += '<h4><i class="fas fa-times-circle"></i> Trạng thái:</h4>'
        html += '<p class="match-fail">Xác thực thất bại</p>'

        if verifyResult.get('msg'):
            html += '<h4><i class="fas fa-exclamation-triangle"></i> Lỗi:</h4>'
            html += f'<p>{verifyResult["msg"]}</p>'

    # Handle legacy format (backward compatibility)
    if verifyResult.get('match_score') is not None:
        html += '<h4><i class="fas fa-percentage"></i> Điểm khớp:</h4>'
        html += f'<p class="match-score">{(verifyResult["match_score"] * 100):.2f}%</p>'

    if verifyResult.get('is_match') is not None:
        matchStatus = "Khớp" if verifyResult['is_match'] else "Không khớp"
        matchClass = "match-success" if verifyResult['is_match'] else "match-fail"
        html += '<h4><i class="fas fa-check-circle"></i> Kết quả:</h4>'
        html += f'<p class="{matchClass}">{matchStatus}</p>'

    if verifyResult.get('confidence') is not None:
        html += '<h4><i class="fas fa-chart-line"></i> Độ tin cậy:</h4>'
        html += f'<p class="confidence">{(verifyResult["confidence"] * 100):.2f}%</p>'

    # Add other fields
    excluded_fields = [
        "status", "same_person", "similarity", "msg",
        "match_score", "is_match", "confidence"
    ]
    for key, value in verifyResult.items():
        if key not in excluded_fields:
            html += f'<h4><i class="fas fa-info-circle"></i> {key}:</h4>'
            html += f'<p>{value}</p>'

    html += "</div>"
    return html

File: backend/app/utils/formatters.py
"""
Data formatting utilities for Laos eKYC Agent
"""

from typing import Dict, Any, Optional
import math


def format_file_size(bytes_size: int) -> str:
    """Format file size in bytes to human readable format"""
    if bytes_size == 0:
        return "0 Bytes"

    k = 1024
    sizes = ["Bytes", "KB", "MB", "GB"]
    i = int(math.floor(math.log(bytes_size) / math.log(k)))

    return f"{math.floor((bytes_size / math.pow(k, i)) * 100) / 100} {sizes[i]}"


def format_scan_result(scan_result: Optional[Dict[str, Any]]) -> str:
    """Format OCR scan result to Markdown"""
    if not scan_result:
        return "ບໍ່ມີຂໍ້ມູນການສະແກນ"

    # Check if it's Lao CCCD format
    if scan_result.get("document_type") == "lao_cccd":
        return format_lao_cccd_result(scan_result)

    # Generic format
    md = "**ຜົນການສະແກນ**\n\n"

    if scan_result.get("text"):
        md += f"**ຂໍ້ຄວາມທີ່ດຶງໄດ້:**\n{scan_result['text']}\n\n"

    if scan_result.get("document_type"):
        md += f"**ປະເພດເອກະສານ:** {scan_result['document_type']}\n"

    if scan_result.get("display_name"):
        md += f"**ຊື່ສະແດງ:** {scan_result['display_name']}\n"

    # Add other fields
    excluded_fields = ["text", "document_type", "display_name", "img_base64", "raw_data", "timestamp"]
    for key, value in scan_result.items():
        if key not in excluded_fields and value:
            md += f"**{key}:** {value}\n"

    return md


def format_lao_cccd_result(ocr_result: Dict[str, Any]) -> str:
    """Format OCR result for Lao CCCD to clean Markdown display"""
    if not ocr_result:
        return "ບໍ່ມີຂໍ້ມູນບັດປະຈຳຕົວ"

    md = "### ບັດປະຈຳຕົວປະຊາຊົນລາວ\n\n"

    # Personal information
    fields = ocr_result.get("fields", {})
    if fields:
        md += "**ຂໍ້ມູນສ່ວນຕົວ**\n"

        # ID number
        if fields.get("id_number"):
            md += f"- **ເລກບັດ:** {fields['id_number']}\n"

        # Full name
        if fields.get("fullname"):
            md += f"- **ຊື່ ແລະ ນາມສະກຸນ:** {fields['fullname']}\n"

        # Date of birth
        if fields.get("dob"):
            md += f"- **ວັນເດືອນປີເກີດ:** {fields['dob']}\n"

        # Nationality
        if fields.get("nationality"):
            md += f"- **ສັນຊາດ:** {fields['nationality']}\n"

        # Ethnicity
        if fields.get("ethnicity"):
            md += f"- **ຊົນເຜົ່າ:** {fields['ethnicity']}\n"

        md += "\n"

        # Address
        address = fields.get("address", {})
        if address:
            md += "**ທີ່ຢູ່**\n"

            if address.get("address"):
                md += f"- **ທີ່ຢູ່:** {address['address']}\n"

            children = address.get("childrent", {})
            if children:
                if children.get("address_village"):
                    md += f"- **ບ້ານ:** {children['address_village']}\n"

                if children.get("address_district"):
                    md += f"- **ເມືອງ:** {children['address_district']}\n"

                if children.get("address_province"):
                    md += f"- **ແຂວງ:** {children['address_province']}\n"

            md += "\n"

        # Dates
        md += "**ຂໍ້ມູນວັນທີ**\n"
        if fields.get("issue_date"):
            md += f"- **ວັນທີອອກບັດ:** {fields['issue_date']}\n"

        if fields.get("expiry_date"):
            md += f"- **ວັນທີໝົດອາຍຸ:** {fields['expiry_date']}\n"

    return md


def format_verify_result(verify_result: Optional[Dict[str, Any]]) -> str:
    """Format face verification result to clean Markdown"""
    if not verify_result:
        return "ບໍ່ມີຂໍ້ມູນການຢັ້ງຢືນ"

    md = "### ຜົນການຢັ້ງຢືນໃບໜ້າ\n\n"

    same_person = verify_result.get("same_person")
    similarity = verify_result.get("similarity")

    if same_person is not None:
        if same_person:
            md += "**ການຢັ້ງຢືນສຳເລັດ!**\n"
            md += "ໃບໜ້າໃນບັດ ແລະ ຮູບຖ່າຍກົງກັນ.\n\n"

            if similarity is not None:
                similarity_percent = (similarity + 1) / 2 * 100
                md += f"- **ຄວາມຄ້າຍຄືກັນ:** {similarity_percent:.1f}%\n"
        else:
            md += "**ການຢັ້ງຢືນລົ້ມເຫລວ**\n"
            md += "ໃບໜ້າໃນບັດ ແລະ ຮູບຖ່າຍບໍ່ກົງກັນ.\n\n"

            if similarity is not None:
                similarity_percent = (similarity + 1) / 2 * 100
                md += f"- **ຄວາມຄ້າຍຄືກັນ:** {similarity_percent:.1f}%\n"
    else:
        md += "**ເກີດຂໍ້ຜິດພາດ**\n"
        md += "ເກີດຂໍ້ຜິດພາດໃນການຢັ້ງຢືນໃບໜ້າ.\n"

    return md
